### PR TITLE
Backport JMS work for Metrics from master 

### PR DESCRIPTION
**Pull request by matzew**
AGPUSH-2072

Refactor metrics collector to be deterministic and complete

:nail_care: it's mainly porting JMS code, and updates to 1.1.x

### CODE REVIEW

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/util/JmsClient.java [hunk 1]
```diff
@@ -207,8 +231,30 @@ public JmsSender inTransaction() {
...
         /**
          * Sets the property that can be later used to query message by selector.
+         *
+         * @param name of property
+         * @param value of property
+         * @return this sender object
          */
         public JmsSender withProperty(String name, String value) {
+            if (value == null) {
+                throw new NullPointerException("property value");
```
Comment (josemigallas) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/util/JmsClient.java]: The error message is a bit unclear, what do you think about `Parameter 'value' can't be null` or `Property value can't be null`. I'm aware it is already a NullPointerException and may look redundant but still, the message reads awkward.

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/configuration/TestSenderConfigurationProducer.java [hunk 2]
```diff
@@ -59,4 +59,16 @@ public void test() {
             System.clearProperty("aerogear.android.batchSize");
         }
     }
+
+    @Test
+    public void testIOS() {
```
Comment (josemigallas) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/configuration/TestSenderConfigurationProducer.java]: What feature are they testing, actually? Could the names be more descriptive?

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/TestTriggerMetricCollectionDeduplication.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    @Resource(mappedName = "java:/queue/TriggerMetricCollectionQueue")
+    private Queue triggerMetricCollectionQueue;
+
+    private static final ConcurrentLinkedQueue<TriggerMetricCollectionEvent> receivedTriggers = new ConcurrentLinkedQueue<TriggerMetricCollectionEvent>();
+    private static final CountDownLatch firstMessageLatch = new CountDownLatch(1);
+    private static String messageId;
+
+    @Test(timeout = 5000)
+    public void testDeduplication() throws InterruptedException {
```
Comment (josemigallas) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/TestTriggerMetricCollectionDeduplication.java]: What kind of duplication is it testing? Is it asserting that metrics _are_ or _aren't_ duplicated?

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/TestTriggerMetricCollectionRedelivery.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+    @Deployment
+    public static WebArchive archive() {
+        return UnifiedPushArchive.forTestClass(TestTriggerMetricCollectionRedelivery.class)
+                .withMessaging()
+                .withMessageDrivenBeans()
+                .as(WebArchive.class);
+    }
+
+    @Test(timeout = 6000)
+    public void testTransactedRedelivery() throws InterruptedException {
```
Comment (josemigallas) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/TestTriggerMetricCollectionRedelivery.java]: What and where is the assertion here?

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/test/archive/UnifiedPushArchiveImpl.java [hunk 5]
```diff
@@ -65,9 +73,18 @@ public UnifiedPushArchive withMessaging() {
...
+            .addAsWebInfResource(new File("../../servers/ups-wildfly/src/main/webapp/WEB-INF/hornetq-jms.xml"));
+    }
+
+    @Override
+    public UnifiedPushArchive withMessageDrivenBeans() {
+        return addClasses(AbstractJMSMessageListener.class)
+                .addClasses(MessageHolderWithVariantsConsumer.class, MessageHolderWithVariantsProducer.class)
+                .addClasses(MessageHolderWithTokensConsumer.class, MessageHolderWithTokensProducer.class)
+                .addClasses(TriggerMetricCollectionConsumer.class)
+                .addAsWebInfResource(new File("../../servers/ups-wildfly/src/main/webapp/WEB-INF/jboss-ejb3.xml"), "jboss-ejb3.xml");
```
Comment (josemigallas) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/test/archive/UnifiedPushArchiveImpl.java]: These paths and file names should be extracted as constants, specially if they are reused in more places.

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/MetricsCollector.java [hunk 6]
```diff
@@ -65,18 +74,52 @@
...
-        // AGPUSH-1585:
-        // using a combined key of variant ID and PushMessageInformation ID, to not limit different push requests on the queue just to the variant
-        // TODO: improve name and/or implementation if this.
-        final String variantPushMessageID = variantMetricInformation.getVariantID() + ":" + pushMessageInformation.getId();
+        metricsService.updatePushMessageInformation(pushMessageInformation);
+    }
 
-        int loadedBatches = countLoadedBatches(variantPushMessageID);
+    private void receiveVariantMetricsRemainingInQueues(PushMessageInformation pushMessageInformation) throws JMSException {
+        while (true) {
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/MetricsCollector.java]: I'd hate to change this.. it's working fine on MASTER.

However,I am fine in doing a fix for that on master. with a new AGPUSH JIRA

File: push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/MetricsCollector.java [hunk 7]
```diff
@@ -92,32 +135,22 @@ public void collectMetrics(@Observes @Dequeue VariantMetricInformation variantMe
...
+        while (receiveBatchLoadedEvent(variantMetricInformation) != null) {
             loadedBatches += 1;
         }
         return loadedBatches;
     }
 
-    private boolean areAllBatchesLoaded(String variantID) {
-        return receiveInTransactionNoWait(allBatchesLoaded, "variantID", variantID) != null;
+    private boolean areAllBatchesLoaded(VariantMetricInformation variantMetricInformation) {
+        if (areIntegersEqual(variantMetricInformation.getTotalBatches(), variantMetricInformation.getServedBatches())) {
```
Comment (matzew) [inline push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/MetricsCollector.java]: I'd like to keep that too. but we could change it on master

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/configuration/TestSenderConfigurationProducer.java (hunk 2 above)
Comment (matzew) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/configuration/TestSenderConfigurationProducer.java]: the batch size, for iOS

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/TestTriggerMetricCollectionDeduplication.java (hunk 3 above)
Comment (matzew) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/TestTriggerMetricCollectionDeduplication.java]: event handling w/ CDI and JMS bridge

File: push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/TestTriggerMetricCollectionRedelivery.java (hunk 4 above)
Comment (matzew) [inline push/sender/src/test/java/org/jboss/aerogear/unifiedpush/message/jms/TestTriggerMetricCollectionRedelivery.java]: I think that the event is processed correctly inside the container.  Not 100% sure. The person that wrote the code is gone

**Pull request merged**